{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "source": [
    "# Task Desription\n",
    "\n",
    "Please read first the following abstract\n",
    "\n",
    "## Abstract\n",
    "                \n",
    "RNA-Seq reveals an unprecedented complexity of the neuroblastoma transcriptome and is suitable for clinical endpoint prediction [ microarray ]\n",
    "\n",
    "### Experiment Description  \n",
    "\n",
    "We generated gene expression profiles from 498 primary neuroblastomas using RNA-Seq and microarrays. We sought to systematically evaluate the capability of RNA deep-sequencing (RNA-Seq)-based classification for clinical endpoint prediction in comparison to microarray-based ones. The neuroblastoma cohort was randomly divided into training and validation sets (**Please note:** <em>in the following we refer to this validation set as test set</em>), and 360 predictive models on six clinical endpoints were generated and evaluated. While prediction performances did not differ considerably between the two technical platforms, the RNA-Seq data processing pipelines, or feature levels (i.e., gene, transcript, and exon junction levels), RNA-Seq models based on the AceView database performed best on most endpoints. Collectively, our study reveals an unprecedented complexity of the neuroblastoma transcriptome, and provides guidelines for the development of gene expression-based predictive classifiers using high-throughput technologies.  Sample clinical characteristics definitions:  \n",
    "\n",
    "* sex: \n",
    "    <ul>\n",
    "    <li>M = male</li>\n",
    "    <li>F = female</li>\n",
    "    </ul>\n",
    "    \n",
    "* age at diagnosis: The age in days at diagnosis \n",
    "    <ul>\n",
    "    <li>integer</li>\n",
    "    </ul>\n",
    "\n",
    "* high risk: Clinically considered as high-risk neuroblastoma\n",
    "    <ul>\n",
    "    <li>yes = 1</li>\n",
    "    <li>no = 0</li>\n",
    "    </ul>\n",
    "\n",
    "\n",
    "* INSS stage: Disease stage according to International Neuroblastoma Staging System ([INSS](https://www.cancer.org/cancer/neuroblastoma/detection-diagnosis-staging/staging.html)) \n",
    "    <ul>\n",
    "    <li>1</li>\n",
    "    <li>2</li>\n",
    "    <li>3</li>\n",
    "    <li>4</li>\n",
    "    <li>4S</li>\n",
    "    </ul>\n",
    "\n",
    "\n",
    "* progression: Occurrence of a tumor progression event\n",
    "    <ul>\n",
    "    <li>yes = 1</li>\n",
    "    <li>no = 0</li>\n",
    "    </ul>\n",
    "\n",
    "\n",
    "\n",
    "* death from disease: Occurrence of death from the disease (yes=1; no=0) \n",
    "    <ul>\n",
    "    <li>yes = 1</li>\n",
    "    <li>no = 0</li>\n",
    "    </ul>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Gene expression of 498 neuroblastoma samples was quantified by RNA sequencing as well as by microarray analyses in order to understand the neuroblastoma transcriptome and predict clinical endpoints. \n",
    "\n",
    "\n",
    "## Task\n",
    "\n",
    "The task is to predict the missing values in the validation set (from here on called test set). Do this either with the RNASeq or the Microarray data, or potentially both?\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code\n",
    "\n",
    "To make your start a bit easier, here is a small notebook reading the data in. It finishes with a function enabling you to save your predictions for submission. \n",
    "\n",
    "#### from here, the code starts\n",
    "\n",
    "First some imports "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt # plotting and visulisation\n",
    "import seaborn as sns # nicer (easier) visualisation\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "# for saving\n",
    "import os,os.path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting  up directory and filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '..{}data'.format(os.path.sep)\n",
    "\n",
    "fn_fpkm             = 'log2FPKM.tsv'\n",
    "fn_patient_info     = 'patientInfo.tsv'\n",
    "fn_prop_intensities = 'allProbIntensities.tsv'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the RNAs-Seq data\n",
    "\n",
    "This part already sets the indeces in the DataFrame. Please feel free to change as required. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID</th>\n",
       "      <th>NB001</th>\n",
       "      <th>NB002</th>\n",
       "      <th>NB003</th>\n",
       "      <th>NB004</th>\n",
       "      <th>NB005</th>\n",
       "      <th>NB006</th>\n",
       "      <th>NB007</th>\n",
       "      <th>NB008</th>\n",
       "      <th>NB009</th>\n",
       "      <th>NB010</th>\n",
       "      <th>...</th>\n",
       "      <th>NB489</th>\n",
       "      <th>NB490</th>\n",
       "      <th>NB491</th>\n",
       "      <th>NB492</th>\n",
       "      <th>NB493</th>\n",
       "      <th>NB494</th>\n",
       "      <th>NB495</th>\n",
       "      <th>NB496</th>\n",
       "      <th>NB497</th>\n",
       "      <th>NB498</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1/2-SBSRNA4</td>\n",
       "      <td>0.834381</td>\n",
       "      <td>0.743094</td>\n",
       "      <td>0.909414</td>\n",
       "      <td>0.795775</td>\n",
       "      <td>0.905540</td>\n",
       "      <td>0.869154</td>\n",
       "      <td>1.811352</td>\n",
       "      <td>0.599240</td>\n",
       "      <td>0.981855</td>\n",
       "      <td>1.066399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997977</td>\n",
       "      <td>1.003559</td>\n",
       "      <td>0.842437</td>\n",
       "      <td>1.057873</td>\n",
       "      <td>0.805515</td>\n",
       "      <td>0.491331</td>\n",
       "      <td>0.868249</td>\n",
       "      <td>0.911379</td>\n",
       "      <td>0.660139</td>\n",
       "      <td>1.152988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.910053</td>\n",
       "      <td>0.941996</td>\n",
       "      <td>1.950857</td>\n",
       "      <td>1.989477</td>\n",
       "      <td>1.942946</td>\n",
       "      <td>1.927608</td>\n",
       "      <td>1.617745</td>\n",
       "      <td>2.161291</td>\n",
       "      <td>1.436439</td>\n",
       "      <td>2.159797</td>\n",
       "      <td>...</td>\n",
       "      <td>2.336929</td>\n",
       "      <td>2.836360</td>\n",
       "      <td>1.205317</td>\n",
       "      <td>2.439868</td>\n",
       "      <td>1.649027</td>\n",
       "      <td>1.451425</td>\n",
       "      <td>1.493852</td>\n",
       "      <td>1.641241</td>\n",
       "      <td>1.994978</td>\n",
       "      <td>1.289534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>1.453191</td>\n",
       "      <td>0.640614</td>\n",
       "      <td>1.156765</td>\n",
       "      <td>1.525277</td>\n",
       "      <td>1.365043</td>\n",
       "      <td>0.899212</td>\n",
       "      <td>1.304178</td>\n",
       "      <td>1.189205</td>\n",
       "      <td>0.771248</td>\n",
       "      <td>1.114787</td>\n",
       "      <td>...</td>\n",
       "      <td>1.182908</td>\n",
       "      <td>1.367371</td>\n",
       "      <td>0.643751</td>\n",
       "      <td>1.096815</td>\n",
       "      <td>0.925425</td>\n",
       "      <td>0.933275</td>\n",
       "      <td>1.208723</td>\n",
       "      <td>0.904511</td>\n",
       "      <td>1.529221</td>\n",
       "      <td>1.102866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A1CF</td>\n",
       "      <td>0.005102</td>\n",
       "      <td>0.005902</td>\n",
       "      <td>0.005192</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025347</td>\n",
       "      <td>0.005682</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024298</td>\n",
       "      <td>0.007295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006678</td>\n",
       "      <td>0.005746</td>\n",
       "      <td>0.004998</td>\n",
       "      <td>0.004853</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.022780</td>\n",
       "      <td>0.018720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>A2LD1</td>\n",
       "      <td>0.580151</td>\n",
       "      <td>0.738233</td>\n",
       "      <td>0.927667</td>\n",
       "      <td>0.936497</td>\n",
       "      <td>0.924853</td>\n",
       "      <td>0.739038</td>\n",
       "      <td>1.018705</td>\n",
       "      <td>0.546324</td>\n",
       "      <td>0.666877</td>\n",
       "      <td>0.865850</td>\n",
       "      <td>...</td>\n",
       "      <td>0.673627</td>\n",
       "      <td>1.401265</td>\n",
       "      <td>0.837443</td>\n",
       "      <td>0.939849</td>\n",
       "      <td>0.743496</td>\n",
       "      <td>0.957837</td>\n",
       "      <td>0.812093</td>\n",
       "      <td>0.488748</td>\n",
       "      <td>1.068072</td>\n",
       "      <td>0.782887</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 498 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID              NB001     NB002     NB003     NB004     NB005     NB006  \\\n",
       "gene_id                                                                   \n",
       "1/2-SBSRNA4  0.834381  0.743094  0.909414  0.795775  0.905540  0.869154   \n",
       "A1BG         1.910053  0.941996  1.950857  1.989477  1.942946  1.927608   \n",
       "A1BG-AS1     1.453191  0.640614  1.156765  1.525277  1.365043  0.899212   \n",
       "A1CF         0.005102  0.005902  0.005192  0.000000  0.025347  0.005682   \n",
       "A2LD1        0.580151  0.738233  0.927667  0.936497  0.924853  0.739038   \n",
       "\n",
       "ID              NB007     NB008     NB009     NB010  ...     NB489     NB490  \\\n",
       "gene_id                                              ...                       \n",
       "1/2-SBSRNA4  1.811352  0.599240  0.981855  1.066399  ...  0.997977  1.003559   \n",
       "A1BG         1.617745  2.161291  1.436439  2.159797  ...  2.336929  2.836360   \n",
       "A1BG-AS1     1.304178  1.189205  0.771248  1.114787  ...  1.182908  1.367371   \n",
       "A1CF         0.000000  0.000000  0.021880  0.000000  ...  0.024298  0.007295   \n",
       "A2LD1        1.018705  0.546324  0.666877  0.865850  ...  0.673627  1.401265   \n",
       "\n",
       "ID              NB491     NB492     NB493     NB494     NB495     NB496  \\\n",
       "gene_id                                                                   \n",
       "1/2-SBSRNA4  0.842437  1.057873  0.805515  0.491331  0.868249  0.911379   \n",
       "A1BG         1.205317  2.439868  1.649027  1.451425  1.493852  1.641241   \n",
       "A1BG-AS1     0.643751  1.096815  0.925425  0.933275  1.208723  0.904511   \n",
       "A1CF         0.000000  0.006678  0.005746  0.004998  0.004853  0.000000   \n",
       "A2LD1        0.837443  0.939849  0.743496  0.957837  0.812093  0.488748   \n",
       "\n",
       "ID              NB497     NB498  \n",
       "gene_id                          \n",
       "1/2-SBSRNA4  0.660139  1.152988  \n",
       "A1BG         1.994978  1.289534  \n",
       "A1BG-AS1     1.529221  1.102866  \n",
       "A1CF         0.022780  0.018720  \n",
       "A2LD1        1.068072  0.782887  \n",
       "\n",
       "[5 rows x 498 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_fpkm = pd.read_csv('{}/{}'.format(data_dir,fn_fpkm),sep='\\t',).rename({'00gene_id':'gene_id'},axis=1)\n",
    "df_fpkm = df_fpkm.set_index(['gene_id'])\n",
    "df_fpkm.columns.name = 'ID'\n",
    "\n",
    "df_fpkm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the Microarray data\n",
    "\n",
    "This part already sets the indeces in the DataFrame. Please feel free to change as required. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID</th>\n",
       "      <th>GeneSymbols</th>\n",
       "      <th>NB001</th>\n",
       "      <th>NB002</th>\n",
       "      <th>NB003</th>\n",
       "      <th>NB004</th>\n",
       "      <th>NB005</th>\n",
       "      <th>NB006</th>\n",
       "      <th>NB007</th>\n",
       "      <th>NB008</th>\n",
       "      <th>NB009</th>\n",
       "      <th>...</th>\n",
       "      <th>NB489</th>\n",
       "      <th>NB490</th>\n",
       "      <th>NB491</th>\n",
       "      <th>NB492</th>\n",
       "      <th>NB493</th>\n",
       "      <th>NB494</th>\n",
       "      <th>NB495</th>\n",
       "      <th>NB496</th>\n",
       "      <th>NB497</th>\n",
       "      <th>NB498</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reporter.Identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>28913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.99</td>\n",
       "      <td>14.94</td>\n",
       "      <td>12.48</td>\n",
       "      <td>14.63</td>\n",
       "      <td>11.89</td>\n",
       "      <td>15.09</td>\n",
       "      <td>13.07</td>\n",
       "      <td>12.00</td>\n",
       "      <td>11.70</td>\n",
       "      <td>...</td>\n",
       "      <td>13.62</td>\n",
       "      <td>13.03</td>\n",
       "      <td>14.98</td>\n",
       "      <td>13.36</td>\n",
       "      <td>13.90</td>\n",
       "      <td>13.00</td>\n",
       "      <td>13.79</td>\n",
       "      <td>14.70</td>\n",
       "      <td>14.03</td>\n",
       "      <td>12.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27262</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.20</td>\n",
       "      <td>10.41</td>\n",
       "      <td>9.27</td>\n",
       "      <td>8.83</td>\n",
       "      <td>7.97</td>\n",
       "      <td>10.33</td>\n",
       "      <td>9.62</td>\n",
       "      <td>8.72</td>\n",
       "      <td>9.36</td>\n",
       "      <td>...</td>\n",
       "      <td>6.26</td>\n",
       "      <td>5.93</td>\n",
       "      <td>6.97</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.62</td>\n",
       "      <td>7.76</td>\n",
       "      <td>8.56</td>\n",
       "      <td>7.74</td>\n",
       "      <td>7.57</td>\n",
       "      <td>7.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.06</td>\n",
       "      <td>5.26</td>\n",
       "      <td>6.45</td>\n",
       "      <td>2.89</td>\n",
       "      <td>2.00</td>\n",
       "      <td>4.80</td>\n",
       "      <td>3.05</td>\n",
       "      <td>6.39</td>\n",
       "      <td>6.43</td>\n",
       "      <td>...</td>\n",
       "      <td>0.93</td>\n",
       "      <td>0.58</td>\n",
       "      <td>1.26</td>\n",
       "      <td>1.38</td>\n",
       "      <td>3.49</td>\n",
       "      <td>2.07</td>\n",
       "      <td>2.26</td>\n",
       "      <td>2.29</td>\n",
       "      <td>2.63</td>\n",
       "      <td>2.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41426</td>\n",
       "      <td>MBL1P</td>\n",
       "      <td>7.45</td>\n",
       "      <td>8.68</td>\n",
       "      <td>6.30</td>\n",
       "      <td>7.30</td>\n",
       "      <td>6.26</td>\n",
       "      <td>7.50</td>\n",
       "      <td>7.43</td>\n",
       "      <td>6.98</td>\n",
       "      <td>8.02</td>\n",
       "      <td>...</td>\n",
       "      <td>5.35</td>\n",
       "      <td>5.57</td>\n",
       "      <td>5.51</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.60</td>\n",
       "      <td>6.38</td>\n",
       "      <td>7.49</td>\n",
       "      <td>6.77</td>\n",
       "      <td>8.13</td>\n",
       "      <td>7.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>37033</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.74</td>\n",
       "      <td>6.63</td>\n",
       "      <td>6.75</td>\n",
       "      <td>6.20</td>\n",
       "      <td>6.57</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.78</td>\n",
       "      <td>4.80</td>\n",
       "      <td>5.15</td>\n",
       "      <td>...</td>\n",
       "      <td>4.58</td>\n",
       "      <td>4.61</td>\n",
       "      <td>3.54</td>\n",
       "      <td>4.55</td>\n",
       "      <td>4.20</td>\n",
       "      <td>7.16</td>\n",
       "      <td>7.07</td>\n",
       "      <td>5.07</td>\n",
       "      <td>6.28</td>\n",
       "      <td>6.34</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 499 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID                  GeneSymbols  NB001  NB002  NB003  NB004  NB005  NB006  \\\n",
       "Reporter.Identifier                                                         \n",
       "28913                       NaN  14.99  14.94  12.48  14.63  11.89  15.09   \n",
       "27262                       NaN   9.20  10.41   9.27   8.83   7.97  10.33   \n",
       "3180                        NaN   5.06   5.26   6.45   2.89   2.00   4.80   \n",
       "41426                     MBL1P   7.45   8.68   6.30   7.30   6.26   7.50   \n",
       "37033                       NaN   6.74   6.63   6.75   6.20   6.57   6.01   \n",
       "\n",
       "ID                   NB007  NB008  NB009  ...  NB489  NB490  NB491  NB492  \\\n",
       "Reporter.Identifier                       ...                               \n",
       "28913                13.07  12.00  11.70  ...  13.62  13.03  14.98  13.36   \n",
       "27262                 9.62   8.72   9.36  ...   6.26   5.93   6.97   5.99   \n",
       "3180                  3.05   6.39   6.43  ...   0.93   0.58   1.26   1.38   \n",
       "41426                 7.43   6.98   8.02  ...   5.35   5.57   5.51   6.30   \n",
       "37033                 6.78   4.80   5.15  ...   4.58   4.61   3.54   4.55   \n",
       "\n",
       "ID                   NB493  NB494  NB495  NB496  NB497  NB498  \n",
       "Reporter.Identifier                                            \n",
       "28913                13.90  13.00  13.79  14.70  14.03  12.31  \n",
       "27262                 7.62   7.76   8.56   7.74   7.57   7.08  \n",
       "3180                  3.49   2.07   2.26   2.29   2.63   2.54  \n",
       "41426                 6.60   6.38   7.49   6.77   8.13   7.11  \n",
       "37033                 4.20   7.16   7.07   5.07   6.28   6.34  \n",
       "\n",
       "[5 rows x 499 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_prop_intensities = pd.read_csv('{}/{}'.format(data_dir,fn_prop_intensities),sep='\\t').set_index(['Reporter.Identifier'])\n",
    "df_prop_intensities.columns.name = 'ID'\n",
    "\n",
    "df_prop_intensities.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the patient factors, including the potential endpoints \n",
    "\n",
    "This part already sets the indeces in the DataFrame. Please feel free to change as required. \n",
    "Please note, that the ```FactorValues``` should have a 1-to-1 correspondence to the factors desc ribed in the abstract. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>FactorValues</th>\n",
       "      <th>FactorValue..Sex.</th>\n",
       "      <th>FactorValue..age.at.diagnosis.</th>\n",
       "      <th>FactorValue..death.from.disease.</th>\n",
       "      <th>FactorValue..high.risk.</th>\n",
       "      <th>FactorValue..inss.stage.</th>\n",
       "      <th>FactorValue..progression.</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>NB498</td>\n",
       "      <td>female</td>\n",
       "      <td>530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB497</td>\n",
       "      <td>female</td>\n",
       "      <td>379</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB496</td>\n",
       "      <td>male</td>\n",
       "      <td>132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB495</td>\n",
       "      <td>male</td>\n",
       "      <td>163</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB494</td>\n",
       "      <td>male</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FactorValues FactorValue..Sex.  FactorValue..age.at.diagnosis.  \\\n",
       "ID                                                               \n",
       "NB498                   female                             530   \n",
       "NB497                   female                             379   \n",
       "NB496                     male                             132   \n",
       "NB495                     male                             163   \n",
       "NB494                     male                              56   \n",
       "\n",
       "FactorValues  FactorValue..death.from.disease.  FactorValue..high.risk.  \\\n",
       "ID                                                                        \n",
       "NB498                                      NaN                      NaN   \n",
       "NB497                                      0.0                      0.0   \n",
       "NB496                                      NaN                      NaN   \n",
       "NB495                                      0.0                      0.0   \n",
       "NB494                                      NaN                      NaN   \n",
       "\n",
       "FactorValues FactorValue..inss.stage.  FactorValue..progression.  \n",
       "ID                                                                \n",
       "NB498                             NaN                        NaN  \n",
       "NB497                               1                        0.0  \n",
       "NB496                             NaN                        NaN  \n",
       "NB495                               1                        0.0  \n",
       "NB494                             NaN                        NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_patient_info = pd.read_csv('{}/{}'.format(data_dir,fn_patient_info),sep='\\t').set_index('ID')\n",
    "df_patient_info.columns.name = 'FactorValues'\n",
    "\n",
    "df_patient_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Divide into training and external testing\n",
    "\n",
    "As you might have already noticed, we removed some of the factor values for some of the patient **ID**s.\n",
    "Every row, where this information is missing indicate a real validation entry. We can use this information and create two separate DataFrames, one for training, one for the validation (testing). \n",
    "\n",
    "The task is to predict the missing values, either with the RNASeq or the Microarray data, or potentially both?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_patient_info_train  = df_patient_info[df_patient_info['FactorValue..death.from.disease.'].notna()]\n",
    "df_patient_info_test   = df_patient_info[df_patient_info['FactorValue..death.from.disease.'].isna()]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>FactorValues</th>\n",
       "      <th>FactorValue..Sex.</th>\n",
       "      <th>FactorValue..age.at.diagnosis.</th>\n",
       "      <th>FactorValue..death.from.disease.</th>\n",
       "      <th>FactorValue..high.risk.</th>\n",
       "      <th>FactorValue..inss.stage.</th>\n",
       "      <th>FactorValue..progression.</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>NB497</td>\n",
       "      <td>female</td>\n",
       "      <td>379</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB495</td>\n",
       "      <td>male</td>\n",
       "      <td>163</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB493</td>\n",
       "      <td>male</td>\n",
       "      <td>190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB491</td>\n",
       "      <td>male</td>\n",
       "      <td>2326</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB489</td>\n",
       "      <td>female</td>\n",
       "      <td>865</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FactorValues FactorValue..Sex.  FactorValue..age.at.diagnosis.  \\\n",
       "ID                                                               \n",
       "NB497                   female                             379   \n",
       "NB495                     male                             163   \n",
       "NB493                     male                             190   \n",
       "NB491                     male                            2326   \n",
       "NB489                   female                             865   \n",
       "\n",
       "FactorValues  FactorValue..death.from.disease.  FactorValue..high.risk.  \\\n",
       "ID                                                                        \n",
       "NB497                                      0.0                      0.0   \n",
       "NB495                                      0.0                      0.0   \n",
       "NB493                                      0.0                      0.0   \n",
       "NB491                                      0.0                      1.0   \n",
       "NB489                                      0.0                      1.0   \n",
       "\n",
       "FactorValues FactorValue..inss.stage.  FactorValue..progression.  \n",
       "ID                                                                \n",
       "NB497                               1                        0.0  \n",
       "NB495                               1                        0.0  \n",
       "NB493                               1                        0.0  \n",
       "NB491                               4                        1.0  \n",
       "NB489                               4                        0.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_patient_info_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>FactorValues</th>\n",
       "      <th>FactorValue..Sex.</th>\n",
       "      <th>FactorValue..age.at.diagnosis.</th>\n",
       "      <th>FactorValue..death.from.disease.</th>\n",
       "      <th>FactorValue..high.risk.</th>\n",
       "      <th>FactorValue..inss.stage.</th>\n",
       "      <th>FactorValue..progression.</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>NB498</td>\n",
       "      <td>female</td>\n",
       "      <td>530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB496</td>\n",
       "      <td>male</td>\n",
       "      <td>132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB494</td>\n",
       "      <td>male</td>\n",
       "      <td>56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB492</td>\n",
       "      <td>male</td>\n",
       "      <td>947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>NB490</td>\n",
       "      <td>female</td>\n",
       "      <td>1759</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "FactorValues FactorValue..Sex.  FactorValue..age.at.diagnosis.  \\\n",
       "ID                                                               \n",
       "NB498                   female                             530   \n",
       "NB496                     male                             132   \n",
       "NB494                     male                              56   \n",
       "NB492                     male                             947   \n",
       "NB490                   female                            1759   \n",
       "\n",
       "FactorValues  FactorValue..death.from.disease.  FactorValue..high.risk.  \\\n",
       "ID                                                                        \n",
       "NB498                                      NaN                      NaN   \n",
       "NB496                                      NaN                      NaN   \n",
       "NB494                                      NaN                      NaN   \n",
       "NB492                                      NaN                      NaN   \n",
       "NB490                                      NaN                      NaN   \n",
       "\n",
       "FactorValues FactorValue..inss.stage.  FactorValue..progression.  \n",
       "ID                                                                \n",
       "NB498                             NaN                        NaN  \n",
       "NB496                             NaN                        NaN  \n",
       "NB494                             NaN                        NaN  \n",
       "NB492                             NaN                        NaN  \n",
       "NB490                             NaN                        NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_patient_info_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis\n",
    "\n",
    "From here on, you will need to use your skills ...\n",
    "\n",
    "### Note:\n",
    "To be clear, there are **multiple** target features/attributes to predict. Say you want to build a model predicting **death from disease** of a patient, your target variable is ```'FactorValue..death.from.disease.'``` and the corresponding target vector <em>y</em> would be as follows:\n",
    "\n",
    "```python\n",
    "y_train = df_patient_info_train['FactorValue..death.from.disease.'].astype(int)\n",
    "```\n",
    "Taking the other data into account (RNASeq or microarray) as ``` X_train``` (you will have to preprocess and split this information yourselves), you could for example build a random forest model:\n",
    "\n",
    "```python\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "random_f_model_death = random_f_model = RandomForestClassifier() \n",
    "random_f_model_death.fit(X_train,y_train)\n",
    "```\n",
    "and predict ```y_test``` using ```X_test```. \n",
    "\n",
    "Obviously, you want to avoid any overfitting and might want to use appropriate validation approaches. \n",
    "\n",
    "Once you have your model and the prediction for the test data, you should be able to fill the ```'FactorValue..death.from.disease.'``` column in the test set. \n",
    "\n",
    "For the submission, please also include the confidence/probability/score for each of the prediction (assume ```1``` to be the value for the positive class). This only applies to the factors:  **high risk**, **progression** and **death from disease** . \n",
    "\n",
    "The other two factors need to be treated differently.  \n",
    "\n",
    "Using this very rough outline, you should be able to predict all factors in the test data. \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
